// tone/build/esm/effect/FeedbackEffect.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Gain } from "../core/context/Gain";
import { readOnly } from "../core/util/Interface";
import { Effect } from "./Effect";
/**
 * FeedbackEffect provides a loop between an audio source and its own output.
 * This is a base-class for feedback effects.
 */
var FeedbackEffect = /** @class */ (function (_super) {
    __extends(FeedbackEffect, _super);
    function FeedbackEffect(options) {
        var _this = _super.call(this, options) || this;
        _this.name = "FeedbackEffect";
        _this._feedbackGain = new Gain({
            context: _this.context,
            gain: options.feedback,
            units: "normalRange",
        });
        _this.feedback = _this._feedbackGain.gain;
        readOnly(_this, "feedback");
        // the feedback loop
        _this.effectReturn.chain(_this._feedbackGain, _this.effectSend);
        return _this;
    }
    FeedbackEffect.getDefaults = function () {
        return Object.assign(Effect.getDefaults(), {
            feedback: 0.125,
        });
    };
    FeedbackEffect.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        this._feedbackGain.dispose();
        this.feedback.dispose();
        return this;
    };
    return FeedbackEffect;
}(Effect));
export { FeedbackEffect };
