// tone/build/esm/component/dynamics/Compressor.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Param } from "../../core/context/Param";
import { ToneAudioNode } from "../../core/context/ToneAudioNode";
import { optionsFromArguments } from "../../core/util/Defaults";
import { readOnly } from "../../core/util/Interface";
/**
 * Compressor is a thin wrapper around the Web Audio
 * [DynamicsCompressorNode](http://webaudio.github.io/web-audio-api/#the-dynamicscompressornode-interface).
 * Compression reduces the volume of loud sounds or amplifies quiet sounds
 * by narrowing or "compressing" an audio signal's dynamic range.
 * Read more on [Wikipedia](https://en.wikipedia.org/wiki/Dynamic_range_compression).
 * @example
 * var comp = new Compressor(-30, 3);
 */
var Compressor = /** @class */ (function (_super) {
    __extends(Compressor, _super);
    function Compressor() {
        var _this = _super.call(this, optionsFromArguments(Compressor.getDefaults(), arguments, ["treshold", "ratio"])) || this;
        _this.name = "Compressor";
        /**
         *  the compressor node
         */
        _this._compressor = _this.context.createDynamicsCompressor();
        _this.input = _this._compressor;
        _this.output = _this._compressor;
        var options = optionsFromArguments(Compressor.getDefaults(), arguments, ["treshold", "ratio"]);
        _this.threshold = new Param({
            context: _this.context,
            convert: false,
            param: _this._compressor.threshold,
            units: "decibels",
            value: options.threshold,
        });
        _this.attack = new Param({
            context: _this.context,
            param: _this._compressor.attack,
            units: "time",
            value: options.attack,
        });
        _this.release = new Param({
            context: _this.context,
            param: _this._compressor.release,
            units: "time",
            value: options.release,
        });
        _this.knee = new Param({
            context: _this.context,
            convert: false,
            param: _this._compressor.knee,
            units: "decibels",
            value: options.knee,
        });
        _this.ratio = new Param({
            context: _this.context,
            convert: false,
            param: _this._compressor.ratio,
            units: "positive",
            value: options.ratio,
        });
        // set the defaults
        readOnly(_this, ["knee", "release", "attack", "ratio", "threshold"]);
        return _this;
    }
    Compressor.getDefaults = function () {
        return Object.assign(ToneAudioNode.getDefaults(), {
            attack: 0.003,
            knee: 30,
            ratio: 12,
            release: 0.25,
            threshold: -24,
        });
    };
    Object.defineProperty(Compressor.prototype, "reduction", {
        /**
         * A read-only decibel value for metering purposes, representing the current amount of gain
         * reduction that the compressor is applying to the signal. If fed no signal the value will be 0 (no gain reduction).
         */
        get: function () {
            return this._compressor.reduction;
        },
        enumerable: true,
        configurable: true
    });
    Compressor.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        this._compressor.disconnect();
        this.attack.dispose();
        this.release.dispose();
        this.threshold.dispose();
        this.ratio.dispose();
        this.knee.dispose();
        return this;
    };
    return Compressor;
}(ToneAudioNode));
export { Compressor };
