// tone/build/esm/core/context/Delay.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Param } from "../context/Param";
import { optionsFromArguments } from "../util/Defaults";
import { readOnly } from "../util/Interface";
import { ToneAudioNode } from "./ToneAudioNode";
/**
 * Wrapper around Web Audio's native [DelayNode](http://webaudio.github.io/web-audio-api/#the-delaynode-interface).
 * @category Core
 */
var Delay = /** @class */ (function (_super) {
    __extends(Delay, _super);
    function Delay() {
        var _this = _super.call(this, optionsFromArguments(Delay.getDefaults(), arguments, ["delayTime", "maxDelay"])) || this;
        _this.name = "Delay";
        var options = optionsFromArguments(Delay.getDefaults(), arguments, ["delayTime", "maxDelay"]);
        var maxDelayInSeconds = _this.toSeconds(options.maxDelay);
        _this.maxDelay = Math.max(maxDelayInSeconds, _this.toSeconds(options.delayTime));
        _this._delayNode = _this.input = _this.output = _this.context.createDelay(maxDelayInSeconds);
        _this.delayTime = new Param({
            context: _this.context,
            param: _this._delayNode.delayTime,
            units: "time",
            value: options.delayTime,
        });
        readOnly(_this, "delayTime");
        return _this;
    }
    Delay.getDefaults = function () {
        return Object.assign(ToneAudioNode.getDefaults(), {
            delayTime: 0,
            maxDelay: 1,
        });
    };
    /**
     *  Clean up.
     */
    Delay.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        this._delayNode.disconnect();
        this.delayTime.dispose();
        return this;
    };
    return Delay;
}(ToneAudioNode));
export { Delay };
