// tone/build/esm/core/context/Gain.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Param } from "../context/Param";
import { optionsFromArguments } from "../util/Defaults";
import { readOnly } from "../util/Interface";
import { ToneAudioNode } from "./ToneAudioNode";
/**
 * A thin wrapper around the Native Web Audio GainNode.
 * The GainNode is a basic building block of the Web Audio
 * API and is useful for routing audio and adjusting gains.
 * @category Core
 */
var Gain = /** @class */ (function (_super) {
    __extends(Gain, _super);
    function Gain() {
        var _this = _super.call(this, optionsFromArguments(Gain.getDefaults(), arguments, ["gain", "units"])) || this;
        _this.name = "Gain";
        /**
         * The wrapped GainNode.
         */
        _this._gainNode = _this.context.createGain();
        // input = output
        _this.input = _this._gainNode;
        _this.output = _this._gainNode;
        var options = optionsFromArguments(Gain.getDefaults(), arguments, ["gain", "units"]);
        _this.gain = new Param({
            context: _this.context,
            convert: options.convert,
            param: _this._gainNode.gain,
            units: options.units,
            value: options.gain,
        });
        readOnly(_this, "gain");
        return _this;
    }
    Gain.getDefaults = function () {
        return Object.assign(ToneAudioNode.getDefaults(), {
            convert: true,
            gain: 1,
            units: "gain",
        });
    };
    /**
     *  Clean up.
     */
    Gain.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        this._gainNode.disconnect();
        this.gain.dispose();
        return this;
    };
    return Gain;
}(ToneAudioNode));
export { Gain };
