// tone/build/esm/instrument/Monophonic.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { FrequencyClass } from "../core/type/Frequency";
import { optionsFromArguments } from "../core/util/Defaults";
import { noOp } from "../core/util/Interface";
import { Instrument } from "../instrument/Instrument";
/**
 * Abstract base class for other monophonic instruments to extend.
 */
var Monophonic = /** @class */ (function (_super) {
    __extends(Monophonic, _super);
    function Monophonic() {
        var _this = _super.call(this, optionsFromArguments(Monophonic.getDefaults(), arguments)) || this;
        var options = optionsFromArguments(Monophonic.getDefaults(), arguments);
        _this.portamento = options.portamento;
        _this.onsilence = options.onsilence;
        return _this;
    }
    Monophonic.getDefaults = function () {
        return Object.assign(Instrument.getDefaults(), {
            detune: 0,
            onsilence: noOp,
            portamento: 0,
        });
    };
    /**
     * Trigger the attack of the note optionally with a given velocity.
     * @param  note The note to trigger.
     * @param  time When the note should start.
     * @param  velocity The velocity scaler determines how "loud" the note will be triggered.
     * @example
     * synth.triggerAttack("C4");
     * @example
     * //trigger the note a half second from now at half velocity
     * synth.triggerAttack("C4", "+0.5", 0.5);
     */
    Monophonic.prototype.triggerAttack = function (note, time, velocity) {
        if (velocity === void 0) { velocity = 1; }
        this.log("triggerAttack", note, time, velocity);
        var seconds = this.toSeconds(time);
        this._triggerEnvelopeAttack(seconds, velocity);
        this.setNote(note, seconds);
        return this;
    };
    /**
     * Trigger the release portion of the envelope
     * @param  time If no time is given, the release happens immediatly
     * @example
     * synth.triggerRelease();
     */
    Monophonic.prototype.triggerRelease = function (time) {
        this.log("triggerRelease", time);
        var seconds = this.toSeconds(time);
        this._triggerEnvelopeRelease(seconds);
        return this;
    };
    /**
     * Get the level of the output at the given time. Measures
     * the envelope(s) value at the time.
     * @param time The time to query the envelope value
     * @return The output level between 0-1
     */
    Monophonic.prototype.getLevelAtTime = function (time) {
        time = this.toSeconds(time);
        return this.envelope.getValueAtTime(time);
    };
    /**
     * Set the note at the given time. If no time is given, the note
     * will set immediately.
     * @param note The note to change to.
     * @param  time The time when the note should be set.
     * @example
     * //change to F#6 in one quarter note from now.
     * synth.setNote("F#6", "+4n");
     * @example
     * //change to Bb4 right now
     * synth.setNote("Bb4");
     */
    Monophonic.prototype.setNote = function (note, time) {
        var computedTime = this.toSeconds(time);
        var computedFrequency = note instanceof FrequencyClass ? note.toFrequency() : note;
        if (this.portamento > 0 && this.getLevelAtTime(computedTime) > 0.05) {
            var portTime = this.toSeconds(this.portamento);
            this.frequency.exponentialRampTo(computedFrequency, portTime, computedTime);
        }
        else {
            this.frequency.setValueAtTime(computedFrequency, computedTime);
        }
        return this;
    };
    return Monophonic;
}(Instrument));
export { Monophonic };
