// tone/build/esm/signal/ToneConstantSource.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { connect } from "../core/Connect";
import { Param } from "../core/context/Param";
import { optionsFromArguments } from "../core/util/Defaults";
import { OneShotSource } from "../source/OneShotSource";
/**
 * Wrapper around the native fire-and-forget ConstantSource.
 * Adds the ability to reschedule the stop method.
 * ***[[Oscillator]] is better for most use-cases***
 */
var ToneConstantSource = /** @class */ (function (_super) {
    __extends(ToneConstantSource, _super);
    function ToneConstantSource() {
        var _this = _super.call(this, optionsFromArguments(ToneConstantSource.getDefaults(), arguments, ["offset"])) || this;
        _this.name = "ToneConstantSource";
        /**
         *  The signal generator
         */
        _this._source = _this.context.createConstantSource();
        var options = optionsFromArguments(ToneConstantSource.getDefaults(), arguments, ["offset"]);
        connect(_this._source, _this._gainNode);
        _this.offset = new Param({
            context: _this.context,
            convert: options.convert,
            param: _this._source.offset,
            units: options.units,
            value: options.offset,
        });
        return _this;
    }
    ToneConstantSource.getDefaults = function () {
        return Object.assign(OneShotSource.getDefaults(), {
            convert: true,
            offset: 1,
            units: "number",
        });
    };
    /**
     * Start the source node at the given time
     * @param  time When to start the source
     */
    ToneConstantSource.prototype.start = function (time) {
        var computedTime = this.toSeconds(time);
        this.log("start", computedTime);
        this._startGain(computedTime);
        this._source.start(computedTime);
        return this;
    };
    ToneConstantSource.prototype._stopSource = function (time) {
        this._source.stop(time);
    };
    ToneConstantSource.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        if (this.state === "started") {
            this.stop();
        }
        this._source.disconnect();
        this.offset.dispose();
        return this;
    };
    return ToneConstantSource;
}(OneShotSource));
export { ToneConstantSource };
